File: backend/app/schemas/workflow_run.py
"""Pydantic schemas for workflow runs and assembled traces."""

from datetime import datetime
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import WorkflowRunStatus, WorkflowRunTrigger
from app.schemas.model_call import ModelCallRead
from app.schemas.tool_call import ToolCallRead


class WorkflowRunCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    project_id: uuid.UUID
    workflow_id: uuid.UUID
    configuration_id: uuid.UUID | None = None
    trigger_type: WorkflowRunTrigger = WorkflowRunTrigger.MANUAL
    external_reference: str | None = Field(default=None, max_length=255)
    input_summary: str | None = None
    metadata_json: dict[str, object] | None = None


class WorkflowRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    workflow_id: uuid.UUID
    configuration_id: uuid.UUID | None
    triggered_by_user_id: uuid.UUID | None
    trigger_type: WorkflowRunTrigger
    external_reference: str | None
    status: WorkflowRunStatus
    input_summary: str | None
    output_summary: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    latency_ms: int | None
    metadata_json: dict[str, object] | None
    created_at: datetime
    updated_at: datetime


class WorkflowRunUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    external_reference: str | None = Field(default=None, max_length=255)
    status: WorkflowRunStatus | None = None
    input_summary: str | None = None
    output_summary: str | None = None
    error_message: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)
    metadata_json: dict[str, object] | None = None


class WorkflowRunCompleteRequest(BaseModel):
    output_summary: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)
    metadata_json: dict[str, object] | None = None


class WorkflowRunFailRequest(BaseModel):
    error_message: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)


class WorkflowRunTraceRead(BaseModel):
    workflow_run: WorkflowRunRead
    model_calls: list[ModelCallRead]
    tool_calls: list[ToolCallRead]

File: frontend/src/types/workflowRun.ts
export type WorkflowRunStatus =
  | "pending"
  | "running"
  | "succeeded"
  | "failed"
  | "cancelled";

export type WorkflowRunTrigger = "manual" | "api" | "simulated" | "scheduled";
export type ModelCallStatus =
  | "pending"
  | "succeeded"
  | "failed"
  | "retried"
  | "fallback_used";
export type ToolCallStatus = "pending" | "succeeded" | "failed";

export interface WorkflowRun {
  id: string;
  project_id: string;
  workflow_id: string;
  configuration_id: string | null;
  triggered_by_user_id: string | null;
  trigger_type: WorkflowRunTrigger;
  external_reference: string | null;
  status: WorkflowRunStatus;
  input_summary: string | null;
  output_summary: string | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  latency_ms: number | null;
  metadata_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface StartWorkflowRunInput {
  project_id: string;
  workflow_id: string;
  configuration_id?: string | null;
  trigger_type?: WorkflowRunTrigger;
  external_reference?: string;
  input_summary?: string;
  metadata_json?: Record<string, unknown>;
}

export interface ModelCallInput {
  sequence_number: number;
  provider: string;
  model_name: string;
  call_type?: string;
  status?: ModelCallStatus;
  prompt_tokens?: number;
  completion_tokens?: number;
  total_tokens?: number;
  latency_ms?: number;
  estimated_cost_usd?: number;
  is_retry?: boolean;
  is_fallback?: boolean;
  request_summary?: string;
  response_summary?: string;
  error_message?: string;
  metadata_json?: Record<string, unknown>;
}

export interface ToolCallInput {
  sequence_number: number;
  tool_name: string;
  status?: ToolCallStatus;
  latency_ms?: number;
  estimated_cost_usd?: number;
  input_summary?: string;
  output_summary?: string;
  error_message?: string;
  metadata_json?: Record<string, unknown>;
}

export interface WorkflowRunTrace {
  workflow_run: WorkflowRun;
  model_calls: Array<ModelCallInput & { id: string; workflow_run_id: string }>;
  tool_calls: Array<ToolCallInput & { id: string; workflow_run_id: string }>;
}
